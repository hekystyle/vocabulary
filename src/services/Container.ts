interface ServiceFactoryApi<TServicesMap> {
  get<U extends keyof TServicesMap>(key: U): TServicesMap[U];
}

type FactoriesMap<TServicesMap> = {
  [K in keyof TServicesMap]: (api: ServiceFactoryApi<TServicesMap>) => TServicesMap[K];
};

export class Container<TServicesMap> {
  constructor(private factories: FactoriesMap<TServicesMap>) {}

  private cache: Partial<{ [K in keyof TServicesMap]: TServicesMap[K] }> = {};

  get<K extends keyof TServicesMap>(key: K): TServicesMap[K] {
    const service = this.cache[key];

    // eslint-disable-next-line @typescript-eslint/no-unnecessary-condition -- false positive
    if (service) return service;

    const factory = this.factories[key];

    const instance = factory({
      get: innerKey => this.get(innerKey),
    });

    this.cache[key] = instance;

    return instance;
  }
}
