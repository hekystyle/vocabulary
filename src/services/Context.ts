import { createContext } from 'react';
import type { Services } from 'services';

/**
 * @private
 */
export const ServicesContext = createContext<Services | undefined>(undefined);
