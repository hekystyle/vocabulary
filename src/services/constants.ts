export const BASE_API_URL = (import.meta.env['VITE_BASE_API_URL'] as string) ?? window.location.origin;
