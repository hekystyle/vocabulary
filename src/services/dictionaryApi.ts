import { ApiResult } from "types";

export class DictionaryApi {
  private controller = new AbortController();

  public abort() {
    this.controller.abort();
  }

  public async word(word: string): Promise<ApiResult> {
    const response = await fetch(
      `https://api.dictionaryapi.dev/api/v2/entries/en_US/${word}`,
      {
        headers: {
          Accept: "application/json",
        },
        signal: this.controller.signal,
      }
    );

    return await response.json();
  }
}
