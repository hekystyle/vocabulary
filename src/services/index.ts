export * from './AppContainer';
export * from './Container';
export * from './Provider';
export * from './Services';
export * from './useServices';
