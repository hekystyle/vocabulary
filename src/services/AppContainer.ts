import { AppDb } from 'services/db';
import { Container } from './Container';
import { Services } from './Services';
import { DelayedTermsRepository, IndexedDbTermsRepository } from './terms';

const DEFAULT_DELAY = 250 as const;

export class AppContainer extends Container<Services> {
  constructor() {
    super({
      db: () => new AppDb('Vocabulary'),
      termsRepository: ({ get }) => {
        const database = get('db');
        const repo = new IndexedDbTermsRepository(database);

        return import.meta.env.DEV
          ? new DelayedTermsRepository(
              repo,
              import.meta.env.VITE_DELAY ? parseInt(import.meta.env.VITE_DELAY, 10) : DEFAULT_DELAY,
            )
          : repo;
      },
    });
  }
}
