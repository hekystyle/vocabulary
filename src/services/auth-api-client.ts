import { z } from 'zod';
import { BASE_API_URL } from './constants';

const profileSchema = z.object({
  email: z.string().email(),
  photoURL: z.string().url().nullish(),
});

export class AuthApiClient {
  // eslint-disable-next-line class-methods-use-this
  async getUserProfile(signal: AbortSignal | undefined) {
    const url = new URL('/api/auth/profile', BASE_API_URL);
    const response = await fetch(url, {
      headers: { Accept: 'application/json' },
      signal,
    });
    const data: unknown = await response.json();
    return profileSchema.parse(data);
  }

  // eslint-disable-next-line class-methods-use-this
  async signOut() {
    const url = new URL('/api/auth/sign-out', BASE_API_URL);
    await fetch(url, { method: 'POST' });
  }
}
