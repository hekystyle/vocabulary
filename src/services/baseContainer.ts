import { AppDb } from '@/services/db';
import { SynchronizerService } from '@/synchronization/service';
import { AuthApiClient } from './auth-api-client';
import { Container } from './Container';
import { HttpTermsApi } from './http-terms-api';
import { LocalDateStorage } from './LocalDateStorage';
import { Services } from './Services';
import { DelayedTermsRepository, IndexedDbTermsRepository } from './terms';

const DEFAULT_DELAY = 250 as const;

export const baseContainer = () =>
  /**
   * Shared container for prod, dev and test environments.
   */
  new Container<Services>({
    env: () => {
      throw new Error('Override env factory');
    },
    db: () => new AppDb('Vocabulary'),
    indexedDbTermsRepository: ({ get }) => new IndexedDbTermsRepository(get('db')),
    termsRepository: ({ get }) => {
      const env = get('env');
      const repo = get('indexedDbTermsRepository');

      return env.DEV
        ? new DelayedTermsRepository(repo, env.VITE_DELAY ? parseInt(env.VITE_DELAY, 10) : DEFAULT_DELAY)
        : repo;
    },
    authApiClient: () => new AuthApiClient(),
    httpTermsApi: () => new HttpTermsApi(),
    synchronizer: ({ get }) =>
      new SynchronizerService(
        get('httpTermsApi'),
        get('indexedDbTermsRepository'),
        get('lastUpdateSyncedFromServerStorage'),
      ),
    lastUpdateSyncedFromServerStorage: () => new LocalDateStorage('lastUpdateSyncedFromServer'),
  });
