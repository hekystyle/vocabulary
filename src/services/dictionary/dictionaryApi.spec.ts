import * as DictionaryApi from 'services/dictionaryApi';
import fs from 'fs';
import path from 'path';

afterAll(() => {
  jest.restoreAllMocks();
});

it('should parse real data', async () => {
  jest.spyOn(window, 'fetch').mockResolvedValue({
    json: () =>
      Promise.resolve(JSON.parse(fs.readFileSync(path.resolve(__dirname, '__fixtures__', 'car.json'), 'utf8'))),
  } as Response);

  const promise = DictionaryApi.fetchWord('hello');

  await expect(promise).resolves.toMatchSnapshot();
});

it('should throw error when response is invalid', async () => {
  jest.spyOn(window, 'fetch').mockResolvedValue({
    json: () => Promise.resolve({ message: 'Word not found' }),
  } as Response);

  await expect(DictionaryApi.fetchWord('hasaki')).rejects.toThrow('Invalid response, expected array, got object');
});
