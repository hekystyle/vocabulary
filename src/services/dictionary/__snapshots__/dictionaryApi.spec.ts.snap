// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should parse real data 1`] = `
Array [
  Word {
    "license": License {
      "name": "CC BY-SA 3.0",
      "url": "https://creativecommons.org/licenses/by-sa/3.0",
    },
    "meanings": Array [
      Meaning {
        "antonyms": Array [],
        "definitions": Array [
          Definition {
            "antonyms": Array [],
            "definition": "A wheeled vehicle that moves independently, with at least three wheels, powered mechanically, steered by a driver and mostly for personal transportation.",
            "example": "She drove her car to the mall.",
            "synonyms": Array [
              "auto",
              "automobile",
              "carriage",
              "motor",
              "motorcar",
              "vehicle",
            ],
          },
          Definition {
            "antonyms": Array [],
            "definition": "A wheeled vehicle, drawn by a horse or other animal; a chariot.",
            "example": "",
            "synonyms": Array [],
          },
          Definition {
            "antonyms": Array [],
            "definition": "An unpowered unit in a railroad train.",
            "example": "The conductor coupled the cars to the locomotive.",
            "synonyms": Array [
              "railcar",
              "wagon",
            ],
          },
          Definition {
            "antonyms": Array [],
            "definition": "An individual vehicle, powered or unpowered, in a multiple unit.",
            "example": "The 11:10 to London was operated by a 4-car diesel multiple unit.",
            "synonyms": Array [],
          },
          Definition {
            "antonyms": Array [],
            "definition": "A passenger-carrying unit in a subway or elevated train, whether powered or not.",
            "example": "From the frontmost car of the subway, he filmed the progress through the tunnel.",
            "synonyms": Array [],
          },
          Definition {
            "antonyms": Array [],
            "definition": "A rough unit of quantity approximating the amount which would fill a railroad car.",
            "example": "We ordered five hundred cars of gypsum.",
            "synonyms": Array [
              "carload",
              "wagonload",
            ],
          },
          Definition {
            "antonyms": Array [],
            "definition": "The moving, load-carrying component of an elevator or other cable-drawn transport mechanism.",
            "example": "Fix the car of the express elevator - the door is sticking.",
            "synonyms": Array [],
          },
          Definition {
            "antonyms": Array [],
            "definition": "The passenger-carrying portion of certain amusement park rides, such as Ferris wheels.",
            "example": "The most exciting part of riding a Ferris wheel is when your car goes over the top.",
            "synonyms": Array [
              "carriage",
            ],
          },
          Definition {
            "antonyms": Array [],
            "definition": "The part of an airship, such as a balloon or dirigible, which houses the passengers and control apparatus.",
            "example": "",
            "synonyms": Array [
              "basket",
              "gondola",
            ],
          },
          Definition {
            "antonyms": Array [],
            "definition": "A sliding fitting that runs along a track.",
            "example": "",
            "synonyms": Array [],
          },
          Definition {
            "antonyms": Array [],
            "definition": "The aggregate of desirable characteristics of a car.",
            "example": "Buy now! You can get more car for your money.",
            "synonyms": Array [],
          },
          Definition {
            "antonyms": Array [],
            "definition": "A floating perforated box for living fish.",
            "example": "",
            "synonyms": Array [],
          },
        ],
        "partOfSpeech": "noun",
        "synonyms": Array [
          "auto",
          "automobile",
          "carriage",
          "motor",
          "motorcar",
          "vehicle",
          "basket",
          "gondola",
          "carload",
          "wagonload",
          "carriage",
          "railcar",
          "wagon",
        ],
      },
    ],
    "phonetic": "/kɑː/",
    "phonetics": Array [
      Phonetic {
        "audio": "https://api.dictionaryapi.dev/media/pronunciations/en/car-uk.mp3",
        "license": License {
          "name": "BY 3.0 US",
          "url": "https://creativecommons.org/licenses/by/3.0/us",
        },
        "sourceUrl": "https://commons.wikimedia.org/w/index.php?curid=9014179",
        "text": "/kɑː/",
      },
      Phonetic {
        "audio": "https://api.dictionaryapi.dev/media/pronunciations/en/car-us.mp3",
        "license": License {
          "name": "BY-SA 3.0",
          "url": "https://creativecommons.org/licenses/by-sa/3.0",
        },
        "sourceUrl": "https://commons.wikimedia.org/w/index.php?curid=424729",
        "text": "/kɑɹ/",
      },
    ],
    "sourceUrls": Array [
      "https://en.wiktionary.org/wiki/car",
    ],
    "word": "car",
  },
  Word {
    "license": License {
      "name": "CC BY-SA 3.0",
      "url": "https://creativecommons.org/licenses/by-sa/3.0",
    },
    "meanings": Array [
      Meaning {
        "antonyms": Array [],
        "definitions": Array [
          Definition {
            "antonyms": Array [],
            "definition": "A turn.",
            "example": "",
            "synonyms": Array [],
          },
        ],
        "partOfSpeech": "noun",
        "synonyms": Array [],
      },
    ],
    "phonetic": "/kɑː/",
    "phonetics": Array [
      Phonetic {
        "audio": "https://api.dictionaryapi.dev/media/pronunciations/en/car-uk.mp3",
        "license": License {
          "name": "BY 3.0 US",
          "url": "https://creativecommons.org/licenses/by/3.0/us",
        },
        "sourceUrl": "https://commons.wikimedia.org/w/index.php?curid=9014179",
        "text": "/kɑː/",
      },
      Phonetic {
        "audio": "https://api.dictionaryapi.dev/media/pronunciations/en/car-us.mp3",
        "license": License {
          "name": "BY-SA 3.0",
          "url": "https://creativecommons.org/licenses/by-sa/3.0",
        },
        "sourceUrl": "https://commons.wikimedia.org/w/index.php?curid=424729",
        "text": "/kɑɹ/",
      },
    ],
    "sourceUrls": Array [
      "https://en.wiktionary.org/wiki/car",
    ],
    "word": "car",
  },
  Word {
    "license": License {
      "name": "CC BY-SA 3.0",
      "url": "https://creativecommons.org/licenses/by-sa/3.0",
    },
    "meanings": Array [
      Meaning {
        "antonyms": Array [
          "cdr",
        ],
        "definitions": Array [
          Definition {
            "antonyms": Array [],
            "definition": "The first part of a cons in LISP. The first element of a list",
            "example": "",
            "synonyms": Array [],
          },
        ],
        "partOfSpeech": "noun",
        "synonyms": Array [],
      },
    ],
    "phonetic": "/kɑː/",
    "phonetics": Array [
      Phonetic {
        "audio": "https://api.dictionaryapi.dev/media/pronunciations/en/car-uk.mp3",
        "license": License {
          "name": "BY 3.0 US",
          "url": "https://creativecommons.org/licenses/by/3.0/us",
        },
        "sourceUrl": "https://commons.wikimedia.org/w/index.php?curid=9014179",
        "text": "/kɑː/",
      },
      Phonetic {
        "audio": "https://api.dictionaryapi.dev/media/pronunciations/en/car-us.mp3",
        "license": License {
          "name": "BY-SA 3.0",
          "url": "https://creativecommons.org/licenses/by-sa/3.0",
        },
        "sourceUrl": "https://commons.wikimedia.org/w/index.php?curid=424729",
        "text": "/kɑɹ/",
      },
    ],
    "sourceUrls": Array [
      "https://en.wiktionary.org/wiki/car",
    ],
    "word": "car",
  },
]
`;
